File: Binary Search/Easy/LET_no74_Search a 2D Matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m = matrix.size(), n = matrix[0].size();
        int left = 0, right = m * n - 1;

        while (left + 1 < right) {
            int mid = left + (right - left) / 2;
            int i = mid / n, j = mid % n;
            if (matrix[i][j] >= target)
                right = mid;
            else 
                left = mid;
        }

        if (matrix[left / n][left % n] == target)
            return true;
        if (matrix[right / n][right % n] == target)
            return true;
        return false;
    }
};

File: Binary Search/Easy/LET_no74_Search a 2D Matrix.py
class Solution:
    """
    @param matrix: matrix, a list of lists of integers
    @param target: An integer
    @return: a boolean, indicate whether matrix contains target
    """

    def searchMatrix(self, matrix, target):
        if not matrix or not matrix[0]:
            return False
        
        cols = len(matrix[0])
        start, end = 0, len(matrix) * cols - 1
        
        while start + 1 < end:
            mid = (start + end) // 2
            i, j = mid // cols, mid % cols
            if target < matrix[i][j]: end = mid
            else: start = mid
        
        i, j = start // cols, start % cols
        if matrix[i][j] == target: return True
        i, j = end // cols, end % cols
        if matrix[i][j] == target: return True
        
        return False
